doc/Style.md

```diff
@@ -0,0 +1,248 @@
+`input` or  `length` could  be UFCS calls,  `input` could be an array,
+`input` could be a member  returning a `struct` which defines `length`, etc..
+
+While template constraints are a remedy to this issue, they require additional
+developer work, which is usually duplicated.
+Finally, templates go through a complex resolution algorithm, and need to be
+instantiated, leading to slow down in compilation.
+
+### Separate / abstract away IO from business code
+
+Separating IO from business code allows to provide mock IO primitives in tests,
+which enables us to simulate an entire network from a `unittest`.
+
+### Balance OO and procedural programming
+
+The traditional D style is very procedural, with seldom use of object.
+In comparison, other close language such as C++ and Java are more OOP-oriented.
+We tend to favor an approach which is closer to C++, althought with reduced
+reliance on constructors / destructors.
+
+
+## Tests
+
+There are four kind of tests we perform: `unittest`, network `unittest`,
+integration tests and testnet tests.
+
+Unittests and network unittests can give us code coverage data,
+which can then be merged to assess that new code is thoroughly tested.
+This is not currently integrated but will be in the future.
+
+### Unittests
+
+Unittests are contained within an `unittest` block.
+They should be simple, short, and independent of other tests.
+They usually test a single function / symbol, or eventually a few working together.
+For example the basic key signature unittest:
+
+```d
+/// Test that `verify`  and `sign` work in concert
+unittest
+{
+    KeyPair kp = KeyPair.random();
+    Signature sign = kp.secret.sign("Hello World".representation);
+    assert(kp.address.verify(sign, "Hello World".representation));
+}
+```
+
+### Network unittests
+
+Network unittests are taking advantage of `agora`'s design to allow testing
+an entire network through `unittest`.
+Those tests live in the `agora.test` package so they can be selectively
+disabled if they grow too large,  or when they don't need to be run.
+
+Following is a template for such a test:
+```d
+
+/*******************************************************************************
+
+    <Test summary>
+
+    <Extended description>
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.test.Name;
+
+version (unittest):
+
+// [...] Any import you need
+import agora.test.Base; // Base functions for the test environment
+
+///
+unittest
+{
+    const NodeCount = 4;
+    auto network = makeTestNetwork!TestNetworkManager(NetworkTopology.Simple, NodeCount);
+    network.start();
+    assert(network.getDiscoveredNodes().length == NodeCount);
+
+   /// Use API as needed
+   foreach (/*PublicKey*/ key, ref /*API*/ node; network.apis)
```

Comment:
Btw, having these `/* */` in the style example is probably a bad idea. It's better to either explicitly use types, or to not use them at all. Because documentation is rarely updated so this might get outdated.

Reply:
The reason it's in `/* */` is because I'm wary of DMD `foreach` type inference errors...